quic/s2n-quic-platform/src/socket/task/rx.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    message::Message,
+    socket::{ring::Producer, task::events},
+};
+use core::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use futures::ready;
+
+pub trait Socket<T: Message> {
+    type Error;
+
+    fn recv(
+        &mut self,
+        cx: &mut Context,
+        entries: &mut [T],
+        events: &mut events::RxEvents,
+    ) -> Result<(), Self::Error>;
+}
+
+pub struct Receiver<T: Message, S: Socket<T>> {
+    ring: Producer<T>,
+    rx: S,
+    pending: u32,
+}
+
+impl<T, S> Receiver<T, S>
+where
+    T: Message + Unpin,
+    S: Socket<T> + Unpin,
+{
+    #[inline]
+    pub fn new(ring: Producer<T>, rx: S) -> Self {
+        Self {
+            ring,
+            rx,
+            pending: 0,
+        }
+    }
+
+    #[inline]
+    fn poll_ring(&mut self, watermark: u32, cx: &mut Context) -> Poll<Option<usize>> {
+        loop {
+            let count = ready!(self.ring.poll_acquire(watermark, cx));
+
+            // if the number of items increased since last time then yield
+            if count > self.pending {
+                return Some(self.pending as usize).into();
+            }
+
+            // if we didn't get any items we can release the ones that we currently have and poll
+            // again
+            self.release();
+        }
+    }
+
+    #[inline]
+    fn release(&mut self) {
+        let to_release = core::mem::take(&mut self.pending);
+        self.ring.release(to_release);
+    }
+}
+
+impl<T, S> Future for Receiver<T, S>
+where
+    T: Message + Unpin,
+    S: Socket<T> + Unpin,
+{
+    type Output = Option<S::Error>;
+
+    #[inline]
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let this = self.get_mut();
+
+        let mut events = events::RxEvents::default();
+
+        while !events.take_blocked() {
+            let pending = match ready!(this.poll_ring(u32::MAX, cx)) {
```

Comment:
> Really the poll here is asking "did you free up more capacity since we last asked", i.e., when we do this.ring.data()[pending..] is that going to be empty. Or at least that's my mental model?

yeah that's correct. are you suggesting the `poll_ring` function just return `Option<()>` and using `this.pending` directly?